### ISSUE TITLE
OpenCV 3.1 DNN Module

### ISSUE THREAD

**Issue by paolobolettieri** [bytedeco/javacpp-presets#145]
A simple question: do the OpenCV presets include the new OpenCV 3.1 Deep Neural Network (DNN) module?
Thanks,

**Comment by saudet** [comment]
No, but it shouldn't be too hard to add. Would you like to make
contribution?

**Comment by paolobolettieri** [comment]
I could try, but I have to talk to my team leader. In the meanwhile, can you tell me where I can find the JavaCPP docs? The only one I found is on https://github.com/bytedeco/javacpp. Is it enough?

**Comment by saudet** [comment]
I would suggest starting here:
https://github.com/bytedeco/javacpp-presets/wiki/Create-New-Presets
But in this case, we'd just basically need to copy, say, this file:
https://github.com/bytedeco/javacpp-presets/blob/master/opencv/src/main/java/org/bytedeco/javacpp/presets/opencv_optflow.java
Modify the settings for `opencv_dnn`, and it might very well work. If not, we may need to add a couple of `Info`, but that should be it.

**Comment by paolobolettieri** [comment]
Hello, I'm trying to create the dnn presets with the code below, but I'm getting the following error

Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project opencv: Compilation failure
opencv_dnn.java:[558,32] <identifier> expected

The row 558 is the following
public native @ByVal cv::Ptr<Layer> call(@ByRef LayerParams params);

It seems a problem with this typedef in layer.hpp
    typedef Ptr<Layer>(*Constuctor)(LayerParams &params);

Maybe a Info could solve it. Any suggestions?
Thanks.

@Properties(inherit = {opencv_core.class}, value = {
    @Platform(include = {"<opencv2/dnn.hpp>", "<opencv2/dnn/dict.hpp>","<opencv2/dnn/blob.hpp>", "<opencv2/dnn/layer.hpp>", "<opencv2/dnn/dnn.hpp>"}, link = "opencv_dnn@.3.1"),
    @Platform(value = "windows", link = "opencv_dnn310")},
        target = "org.bytedeco.javacpp.opencv_dnn")
public class opencv_dnn implements InfoMapper {
    public void map(InfoMap infoMap) {
    }
}

**Comment by paolobolettieri** [comment]
I removed `"<opencv2/dnn/layer.hpp>"` from _includes_, but now I'm getting the following error:

_jniopencv_dnn.cpp(2533) : error C2664: 'void cv::dnn::Layer::allocate(const std::vector<cv::dnn::Blob *,std::allocator<_Other>> &,std::vector<cv::dnn::Blob,std::allocator<_Ty>> &)': Cannot convert argument 1 from 'VectorAdapter<cv::dnn::Blob,P>' to 'const std::vector<cv::dnn::Blob *,std::allocator<_Other>> &'
        with
        [
            _Other=cv::dnn::Blob *
,            _Ty=cv::dnn::Blob
        ]
        and
        [
            P=cv::dnn::Blob
        ]
        and
        [
            _Other=cv::dnn::Blob *
        ]
       No user-defined-conversion operator available that can perform this conversion, or the operator cannot be call_

**Comment by paolobolettieri** [comment]
Ok, the bug has been fixed, thanks. 
Now I'm getting the following error:

jniopencv_dnn.cpp(2533) : error C2664: 'void cv::dnn::Layer::allocate(const std::vector> &,std::vectorcv::dnn::Blob,std::allocator<_Ty> &)': Cannot convert argument 1 from 'VectorAdaptercv::dnn::Blob,P' to 'const std::vector> &'
with
[
_Other=cv::dnn::Blob *
, _Ty=cv::dnn::Blob
]
and
[
P=cv::dnn::Blob
]
and
[
_Other=cv::dnn::Blob *
]
No user-defined-conversion operator available that can perform this conversion, or the operator cannot be call

**Comment by saudet** [comment]
Like I said in a comment that seems to have been deleted for some reason, we're going to need to define a `BlobVector` like done for `caffe` here:
https://github.com/bytedeco/javacpp-presets/blob/master/caffe/src/main/java/org/bytedeco/javacpp/presets/caffe.java#L137

**Comment by paolobolettieri** [comment]
Done! But unfortunately there is a new Problem :-(

jniopencv_dnn.obj : error LNK2001: unresolved external symbol "public: void __cdecl cv::dnn::Blob::fill(class cv::_InputArray const &)" (?fill@Blob@dnn@cv@@QEAAXAEBV_InputArray@3@@Z) unresolved
jniopencv_dnn.obj : error LNK2001: unresolved external symbol "class cv::dnn::Blob __cdecl cv::dnn::readTorchBlob(class cv::String const &,bool)" (?readTorchBlob@dnn@cv@@YA?AVBlob@12@AEBVString@2@_N@Z) unresolved
jniopencv_dnn.obj : error LNK2001: unresolved external symbol "public: void __cdecl cv::dnn::Net::setParam(struct cv::dnn::DictValue,int,class cv::dnn::Blob const &)" (?setParam@Net@dnn@cv@@QEAAXUDictValue@23@HAEBVBlob@23@@Z) unresolved
jniopencv_dnn.obj : error LNK2001: unresolved external symbol "public: void __cdecl cv::dnn::Net::forwardOpt(struct cv::dnn::DictValue)" (?forwardOpt@Net@dnn@cv@@QEAAXUDictValue@23@@Z) unresolved
jniopencv_dnn.obj : error LNK2001: unresolved external symbol "public: void __cdecl cv::dnn::Net::forward(struct cv::dnn::DictValue,struct cv::dnn::DictValue)" (?forward@Net@dnn@cv@@QEAAXUDictValue@23@0@Z) unresolved
D:\demo\javacv\javacpp-presets\opencv\target\classes\org\bytedeco\javacpp\windows-x86_64\jniopencv_dnn.dll : fatal error LNK1120: 5 unresolved extern

**Comment by saudet** [comment]
I don't have such problems on Linux. It looks like it doesn't support
Windows yet... Could you ask the authors about that? Thanks!

**Comment by saudet** [comment]
Ah, no, this isn't a problem with Windows. These functions are simply not defined:
```
libjniopencv_dnn.so: undefined reference to `cv::dnn::Net::forwardOpt(cv::dnn::DictValue)'
libjniopencv_dnn.so: undefined reference to `cv::dnn::Net::setParam(cv::dnn::DictValue, int, cv::dnn::Blob const&)'
libjniopencv_dnn.so: undefined reference to `cv::dnn::readTorchBlob(cv::String const&, bool)'
libjniopencv_dnn.so: undefined reference to `cv::dnn::Net::forward(cv::dnn::DictValue, cv::dnn::DictValue)'
libjniopencv_dnn.so: undefined reference to `cv::dnn::Blob::fill(cv::_InputArray const&)'
```
So, we're going to need to add `new Info(...).skip()` for those to get this working on Windows. Unless the authors are DNN would be so kind to fix their code for us. :)

**Comment by paolobolettieri** [comment]
Great!, I built it.
In the next days I'll check if it works (fine ;-))

**Comment by saudet** [comment]
Great! Please send a pull request when done! Thanks

**Comment by paolobolettieri** [comment]
Hello,
DNN module works fine!!!
I just need a (last) little help. I'm not able to just skip the five non-defined functions.

For example, to skip 
_cv::dnn::Net::forward(cv::dnn::DictValue, cv::dnn::DictValue)_

I set
`infoMap.put(new Info("cv::dnn::Net::forward").skip())`

but it skips all 'forward' functions and not only the needed.
Also, I tried the following one

`infoMap.put(new Info("cv::dnn::Blob::forward(cv::dnn::DictValue, cv::dnn::DictValue)").skip())`

but it doesn't work.
Can you give me a help please?
Thanks

**Comment by saudet** [comment]
Yeah the Parser isn't so great here. Try to remove the space after the
comma. That should work.

**Comment by paolobolettieri** [comment]
Unfortunately it doesn't work :(
I'm only able to skip `fill`
`infoMap.put(new Info("cv::dnn::Blob::fill(cv::InputArray)").skip());`

**Comment by AmlFawzy** [comment]
can i create iris detection using javacv please help?

**Comment by saudet** [comment]
Ok, I've committed some presets that should work. Enjoy!

**Comment by paolobolettieri** [comment]
Thank you!
In the meanwhile, I wrote the JavaCV version of the following OpenCV DNN example

[](http://docs.opencv.org/3.1.0/d5/de7/tutorial_dnn_googlenet.html#gsc.tab=0)

It might be useful to someone. I can commit it on github if you agree on this.

**Comment by saudet** [comment]
That would be great, thanks! If it's simple enough and fits in one Java file + small data, please send a pull request for the samples directory here:
    https://github.com/bytedeco/javacv/tree/master/samples
If you'd like to contribute a sample project instead, please send a pull request here:
    https://github.com/bytedeco/sample-projects
Thank you!

**Comment by saudet** [comment]
Included in the newly released version 1.2. Enjoy!
